ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/LedgerDB/DbChangelog.hs

```diff
@@ -0,0 +1,427 @@
+      }
+
+  shrink dblog = takeWhileJust $ tail (iterate reduce (Just dblog))
+    where
+      reduce (Just (DbChangelogTestSetup (_:ops) dblog')) = Just $ DbChangelogTestSetup ops dblog'
+      reduce _ = Nothing
+      takeWhileJust = catMaybes . takeWhile isJust
+
+instance Arbitrary DbChangelogTestSetupWithRollbacks where
+  arbitrary = do
+    setup <- arbitrary
+    let dblog = resultingDbChangelog setup
+    rollbacks <- chooseInt (0, AS.length (changelogVolatileStates dblog))
+    pure $ DbChangelogTestSetupWithRollbacks
+      { testSetup = setup
+      , rollbacks = rollbacks
+      }
+
+emptyDbChangelogAtSlot :: WithOrigin SlotNo -> DbChangelog TestLedger
+emptyDbChangelogAtSlot slotNo = emptyDbChangeLog (TestLedger ApplyEmptyMK $ pointAtSlot slotNo)
+
+resultingDbChangelog :: DbChangelogTestSetup -> DbChangelog TestLedger
+resultingDbChangelog setup = applyOperations (operations setup) (originalDbChangelog setup)
+
+applyOperations :: (TableStuff l, GetTip (l EmptyMK))
+  => [Operation l] -> DbChangelog l -> DbChangelog l
+applyOperations ops dblog = foldr' apply' dblog ops
+  where apply' (Extend newState) dblog' = extendDbChangelog dblog' newState
+        apply' (Prune sp) dblog'        = pruneVolatilePartDbChangelog sp dblog'
+
+
+{-------------------------------------------------------------------------------
+  Invariants
+-------------------------------------------------------------------------------}
+
+volatileTipAnchorsImmutable :: (GetTip (l EmptyMK), Eq (l EmptyMK)) => DbChangelog l -> Bool
+volatileTipAnchorsImmutable DbChangelog { changelogImmutableStates, changelogVolatileStates } =
+  AS.anchor changelogVolatileStates == AS.headAnchor changelogImmutableStates
+
+immutableAnchored :: DbChangelog TestLedger -> Bool
+immutableAnchored DbChangelog { changelogDiffAnchor, changelogImmutableStates } =
+  changelogDiffAnchor == fmap Point.blockPointSlot point
+  where point = getPoint $ getTip $ unDbChangelogState $ AS.anchor $ changelogImmutableStates
+
+sameNumberOfDiffsAsStates :: DbChangelog TestLedger -> Bool
+sameNumberOfDiffsAsStates dblog = AS.length imm + AS.length vol == lengthSeqUtxoDiff diffs
+  where imm = changelogImmutableStates dblog
+        vol = changelogVolatileStates dblog
+        ApplySeqDiffMK diffs = unTestTables $ changelogDiffs dblog
+
+checkInvariants :: DbChangelog TestLedger -> Bool
+checkInvariants dblog = volatileTipAnchorsImmutable dblog &&
+                        immutableAnchored dblog &&
+                        sameNumberOfDiffsAsStates dblog
+
+
+{-------------------------------------------------------------------------------
+  Properties
+-------------------------------------------------------------------------------}
+
+prop_emptySatisfiesInvariants :: Property
+prop_emptySatisfiesInvariants =
+  property $ checkInvariants (emptyDbChangelogAtSlot Origin)
+
+prop_generatedSatisfiesInvariants :: DbChangelogTestSetup -> Property
```

Comment:
I wonder if we should check the invariant is satisfied after each operation :thinking: